kotlinx-coroutines-core/common/src/CoroutineStart.kt

```diff
@@ -1,74 +1,342 @@
+     *   when already on the main thread and for [Dispatchers.Unconfined]),
+     *   the task is executed immediately in the same thread that invoked the coroutine builder,
+     *   similarly to [UNDISPATCHED].
      *
-     * If the [CoroutineDispatcher] of the coroutine context returns `true` from [CoroutineDispatcher.isDispatchNeeded]
-     * function as most dispatchers do, then the coroutine code is dispatched for execution later, while the code that
-     * invoked the coroutine builder continues execution.
+     * If the coroutine's [Job] is cancelled before it started executing, then it will not start its
+     * execution at all, and will instead complete with an exception.
      *
-     * Note that [Dispatchers.Unconfined] always returns `false` from its [CoroutineDispatcher.isDispatchNeeded]
-     * function, so starting a coroutine with [Dispatchers.Unconfined] by [DEFAULT] is the same as using [UNDISPATCHED].
+     * Comparisons with the other options:
+     * - [LAZY] delays the moment of the initial dispatch until the completion of the coroutine is awaited.
+     * - [ATOMIC] prevents the coroutine from being cancelled before its first suspension point.
+     * - [UNDISPATCHED] always executes the coroutine until the first suspension immediately in the same thread
+     *   (as if [CoroutineDispatcher.isDispatchNeeded] returned `false`),
+     *   and also, like [ATOMIC], it ensures that the coroutine cannot be cancelled before it starts executing.
      *
-     * If coroutine [Job] is cancelled before it even had a chance to start executing, then it will not start its
-     * execution at all, but will complete with an exception.
+     * Examples:
      *
-     * Cancellability of a coroutine at suspension points depends on the particular implementation details of
-     * suspending functions. Use [suspendCancellableCoroutine] to implement cancellable suspending functions.
+     * ```
+     * // Example of starting a new coroutine that goes through a dispatch
+     * runBlocking {
+     *     println("1. About to start a new coroutine.")
+     *     // Dispatch the job to execute later.
+     *     // The parent coroutine's dispatcher is inherited by default.
+     *     // In this case, it's the single thread backing `runBlocking`.
+     *     val job = launch {
+     *         println("3. When the thread is available, we start the coroutine")
+     *     }
+     *     println("2. The thread keeps doing other work after launching the coroutine")
+     * }
+     * ```
+     *
+     * ```
+     * // Example of starting a new coroutine that doesn't go through a dispatch initially
+     * runBlocking {
+     *     println("1. About to start a coroutine not needing a dispatch.")
+     *     // Dispatch the job to execute.
+     *     // `Dispatchers.Unconfined` is explicitly chosen.
+     *     val job = launch(Dispatchers.Unconfined) {
```

Comment:
After you suggested this, I did have to spend quite some time pondering the issue!

Here's the table that summarizes the behavior for all combinations of `CoroutineStart` and `CoroutineDispatcher.isDispatchNeeded`:

| **Start option** | **`isDispatchNeeded`** | **Behavior** |
| -- | -- | -- |
| `DEFAULT` | yes | dispatch |
| `ATOMIC` | yes | dispatch, prohibiting cancellation |
| `LAZY` | yes | dispatch when requested |
| `UNDISPATCHED` | - | start in-place |
| `DEFAULT` | no | start in-place |
| `ATOMIC` | no | start in-place, prohibiting cancellation |
| `LAZY` | no | the requester starts the coroutine in-place |

I struggle to find an internally consistent model of this behavior that doesn't involve enumerating all of these possibilities in some form.

1. We can define the `if (isDispatchNeeded) dispatch() else dispatchToEventLoop()` procedure as some kind of a "standard" dispatch procedure. Then, the table becomes:

| **Start option** | **Behavior** |
| -- | -- |
| `DEFAULT` | standard dispatch |
| `ATOMIC` | standard dispatch, prohibiting cancellation |
| `LAZY` | standard dispatch when requested |
| `UNDISPATCHED` | start in-place |

`UNDISPATCHED` does not fit it, as it breaks the standard dispatch procedure.

2. We can *not* declare that `isDispatchNeeded = false` simply overrides the dispatching behavior to unconditionally use the event loop, as only half of the cases is covered in this manner:

* `LAZY` with `isDispatchNeeded = false` will still wait until someone requests the result.
* `UNDISPATCHED` will not use the event loop in any case, even though the code is executed in-place.

Also, to add to the problem, I didn't find anything in the contract of https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-dispatcher/is-dispatch-needed.html that prohibited side effects, so maybe every `CoroutineStart` entry should either document whether and when it invokes that function or explicitly state that there is this is undefined...

To sum it up, I do not understand the mental model of `CoroutineStart` that does not include an explicit description of how `isDispatchNeeded = false` is handled. Furthermore, https://github.com/Kotlin/api-guidelines/blob/main/team_docs/samples.md#functions-returning-a-result from the doc you linked suggests that, since `isDispatchNeeded` affects the behavior of `CoroutineStart` significantly, it should be included as a sample.

That's my current understanding, though of course, I will concur with any decision regarding this.

Reply:
I see, thanks.

The mental model (here, and, in fact, of any API) is that different concepts are completely orthogonal, and one doesn't have to think about them in total, only separately (otherwise, the learning curve is unreasonable steep).

`isDispatchNeeded` is a dispatcher property that hints at how exactly the coroutine is dispatched. Coroutine start is about the start, and it's completely orthogonal in every case except `UNDISPATCHED` (and that's the place where it's worth mentioning explicitly).

With such separation in mind, there is no need to enumerate every possibility (the same way we don't enumerate how `DEFAULT` + cancellation + `NonCancellable` works and that `ATOMIC` here does not make any difference), only the core concept -- the start.
`UNDISPATCHED` here is worth a special note in its own documentation.

Hope that clarifies the idea! Having that in mind, it may become more obvious why the `Unconfined` example is unnecessary here, and in the third example, one coroutine is enough.

Small side note: `ATOMIC` does not "prohibit" cancellation, it delays it to the first cancellation check (whether it's a suspension point or `ensureActive`).

Reply:
I would assume that the confusion comes from the fact that we mentioned `isDispatchNeeded` in the original documentation to `DEFAULT`. 
I maybe made sense back then (2017!) when everything was experimental and we had no idea what coroutines are, but I wouldn't do it nowadays

Reply:
> Just double-checking: would you agree that this is exactly the "standard dispatch" explanation I've outlined in my previous message?

I'm not @qwwdfsad and can't be certain, but yeah, I do think you're talking about the same thing.

> If there is a centralized place (for example, in the CoroutineDispatcher KDoc) that we can link to to explain what procedure the coroutine goes through on a "standard dispatch,"

To my knowledge, there isn't one. Just a collection of separate function whose documentation, taken as a whole, explains what's going on.

> we can introduce such a place as part of this PR.

That's probably the way to go. I'll give it a try.

Reply:
Done.